### PR TITLE
N°1646 - Enum: logical ordering defined in datamodel (dashlet, list, transition menu...)

### PR DESCRIPTION
**Pull request by accognet**
Order in select list
Before the PR: 
image

In xml, I have change field definition:
image

Result in screen :
image

What I have made:
- Add 2 tags: _<sort_order>_ and _<rank>_
- _sort_order_ can have 3 values: _code_, _rank_ or _label_.  
     - _code_ is the default value - it works like today. the codes are listed in alphabetical order
     - with _rank_, you have to specify rank of all of the values
     - with _label_, the values are sorted in alphabetical order according to the language of the user.
  
The transition buttons are ordered according to the arrival state. First transitions to progress in lifecycle and then transitions to go back in the lifecycle.

### CODE REVIEW

**Comment by jbostoen** [comment]
Would be nice if sort could happen alphabetically (the order may be different for some languages) 😁

I suppose there could be a field property "sorting" with could default to "rank" or "alphabetical" in such case.

**Comment by accognet** [comment]
Yes, if there is no rank, order is alphabetically depending of language

**Comment by Hipska** [comment]
> Yes, if there is no rank, order is alphabetically depending of language

So the default behaviour will change then? As currently it is sorted on the enum value?

File: core/attributedef.class.inc.php [hunk 1]
```diff
@@ -5888,14 +5888,32 @@ public function GetForJSON($value)
...
+				$aLocalizedValues[$sKey] = $this->GetValueLabel($sKey);
+			} else {
+				$aSortedValues[$aKey[1]] = $aKey[0];
+				$bHaveSortedValues = true;
+			}
+		}
+		asort($aLocalizedValues);
+		
+		if ($bHaveSortedValues) {
+			asort($aLocalizedValues);
```
Comment (Hipska) [inline core/attributedef.class.inc.php]: Duplicate sort of array
```suggestion
		asort($aLocalizedValues);
		
		if ($bHaveSortedValues) {
```

**Comment by accognet** [comment]
Yes, , this is the idea.

**Comment by Hipska** [comment]
Okay, but then this PR should also add orderings to the fields `impact`, `priority` and `urgency` on the different`Ticket` subclasses like for example on `UserRequest`.

https://github.com/Combodo/iTop/blob/2907efde4ba53ead6c797acad37d8b5a404ca2fd/datamodels/2.x/itop-request-mgmt/datamodel.itop-request-mgmt.xml#L150-L233

There might be also other cases in the default datamodel that need to be fixed as well..

**Comment by accognet** [comment]
Yes, there is still work on this subject. This is the reason why this PR is on draft .

**Comment by Hipska** [comment]
Perfect, just wanted to make sure!

File: core/attributedef.class.inc.php [hunk 2]
```diff
@@ -5888,14 +5888,31 @@ public function GetForJSON($value)
...
+			if (count($aVal) == 1) {
+				$aLocalizedValues[$sValue] = $this->GetValueLabel($sValue);
+			} else {
+				$aSortedValues[$aVal[1]] = $aVal[0];
+				$bHaveSortedValues = true;
+			}
+		}
+		asort($aLocalizedValues);
+		
+		if ($bHaveSortedValues) {
```
Comment (Hipska) [inline core/attributedef.class.inc.php]: I don't think this boolean is needed:
```suggestion
		if (!empty($aSortedValues)) {
```

File: datamodels/2.x/itop-config-mgmt/datamodel.itop-config-mgmt.xml [hunk 3]
```diff
@@ -315,15 +315,19 @@
...
+              <rank>1</rank>
             </value>
             <value id="implementation">
               <code>implementation</code>
+              <rank>5</rank>
             </value>
             <value id="stock">
               <code>stock</code>
+              <rank>2</rank>
             </value>
```
Comment (Hipska) [inline datamodels/2.x/itop-config-mgmt/datamodel.itop-config-mgmt.xml]: If you want to mimic current ordering, it should be according to order in xml:
```suggestion
            <value id="implementation">
              <code>implementation</code>
              <rank>2</rank>
            </value>
            <value id="stock">
              <code>stock</code>
              <rank>3</rank>
            </value>
```

Comment (accognet) [reply datamodels/2.x/itop-config-mgmt/datamodel.itop-config-mgmt.xml]: I have to validate this ordering with the product team. It's always work in progress.

File: tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.0_to_3.1.expected.xml [hunk 4]
```diff
@@ -62,5 +62,61 @@
...
+          <values>
+            <value id="new">
+              <code>new</code>
+              <rank>1</rank>
+            </value>
+            <value id="assigned">
+              <code>assigned</code>
+              <rank>0</rank>
+            </value>
+          </values>
```
Comment (Hipska) [inline tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.0_to_3.1.expected.xml:78]: This is the expected order:
```suggestion
          <values>
            <value id="new">
              <code>new</code>
              <rank>0</rank>
            </value>
            <value id="assigned">
              <code>assigned</code>
              <rank>1</rank>
            </value>
          </values>
```

File: tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.0_to_3.1.expected.xml [hunk 5]
```diff
@@ -62,5 +62,61 @@
...
+              <rank>1</rank>
+            </value>
+            <value id="resolved">
+              <code>resolved</code>
+              <rank>2</rank>
+            </value>
+            <value id="closed">
+              <code>closed</code>
+              <rank>0</rank>
+            </value>
```
Comment (Hipska) [inline tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.0_to_3.1.expected.xml:96]:
```suggestion
            <value id="ongoing">
              <code>ongoing</code>
              <rank>0</rank>
            </value>
            <value id="resolved">
              <code>resolved</code>
              <rank>1</rank>
            </value>
            <value id="closed">
              <code>closed</code>
              <rank>2</rank>
            </value>
```

File: tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.1_to_3.0.expected.xml [hunk 6]
```diff
@@ -77,6 +77,57 @@
...
+      <fields>
+        <field id="status" xsi:type="AttributeEnum">
+          <always_load_in_tables>true</always_load_in_tables>
+          <values>
+            <value id="new">
+              <code>new</code>
+            </value>
+            <value id="assigned">
+              <code>assigned</code>
+            </value>
```
Comment (Hipska) [inline tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.1_to_3.0.expected.xml:90]: The input suggested `assigned` before `new`.
```suggestion
            <value id="assigned">
              <code> assigned</code>
            </value>
            <value id="new">
              <code>new</code>
            </value>
```

File: tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.1_to_3.0.expected.xml [hunk 7]
```diff
@@ -77,6 +77,57 @@
...
+            <value id="ongoing">
+              <code>ongoing</code>
+            </value>
+            <value id="resolved">
+              <code>resolved</code>
+            </value>
+            <value id="closed">
+              <code>closed</code>
+            </value>
+          </values>
```
Comment (Hipska) [inline tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.1_to_3.0.expected.xml:107]:
```suggestion
          <values>
            <value id="closed">
              <code>closed</code>
            </value>
            <value id="ongoing">
              <code>ongoing</code>
            </value>
            <value id="resolved">
              <code>resolved</code>
            </value>
          </values>
```

**Review by Hipska — CHANGES_REQUESTED** [review]
Shouldn't it be like this? Or am I wrong with how iTop currently orders enums?

File: datamodels/2.x/itop-config-mgmt/datamodel.itop-config-mgmt.xml (hunk 3 above)
Comment (Hipska) [reply datamodels/2.x/itop-config-mgmt/datamodel.itop-config-mgmt.xml]: Sorry, I was wrong. The current sorting is indeed based on code.

**Comment by piRGoif** [comment]
Functional review : 
- Transitions buttons order : nee to check with other people in Combodo (align with business needs)
- Maybe split the PR : enum order solely in this PR, transition buttons in a dedicated new PR

File: setup/itopdesignformat.class.inc.php [hunk 8]
```diff
@@ -1135,6 +1135,10 @@ protected function From31To30($oFactory)
 		// N°2783 Custom zlists
 		$this->RemoveNodeFromXPath("/itop_design/classes//class/presentation/custom_presentations");
 		$this->RemoveNodeFromXPath("/itop_design/meta/presentation/custom_presentations");
+
+		// N°1646 - Enum: logical ordering defined in datamodel (dashlet, list, transition menu...)
+		// - remove rank in Enum
+		$this->RemoveNodeFromXPath("/itop_design/classes//class/fields/field/values/value/rank");
```
Comment (Molkobain) [inline setup/itopdesignformat.class.inc.php:1141]: Missing `sort_order` node removal

File: tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.0_to_3.1.expected.xml
Comment (Molkobain) [inline tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.0_to_3.1.expected.xml:1]: Not necessary as we don't add nodes automatically on XML upgrade

File: tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.0_to_3.1.input.xml
Comment (Molkobain) [inline tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.0_to_3.1.input.xml:1]: Not necessary as we don't add nodes automatically on XML upgrade

File: tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.1_to_3.0.expected.xml
Comment (Molkobain) [inline tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.1_to_3.0.expected.xml:1]: Reduce XML to only what we are testing for better readability / maintenability.
Name class / attribute with explicit name of the test case so we know what we are testing exactly.

File: tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.1_to_3.0.input.xml
Comment (Molkobain) [inline tests/php-unit-tests/unitary-tests/setup/iTopDesignFormat/Convert-samples/3.1_to_3.0.input.xml:1]: Reduce XML to only what we are testing for better readability / maintenability.
Name class / attribute with explicit name of the test case so we know what we are testing exactly.

File: setup/compiler.class.inc.php [hunk 9]
```diff
@@ -2095,14 +2095,30 @@ protected function CompileAttribute(string $sAttType, DOMElement $oField, string
...
 			$aStyledValues = [];
 			foreach ($oValueNodes as $oValue) {
 				// New in 3.0 the format of values changed
 				$sCode = $this->GetMandatoryPropString($oValue, 'code', false);
-				$aValues[] = $sCode;
+				if ($aParameters['sort_order'] === "'rank'") {
+					$iRank = $oValue->GetChildText('rank');
+					if ($iRank) {
+						$aValues[] = $iRank.'::'.$sCode;
+					} else {
```
Comment (Molkobain) [inline setup/compiler.class.inc.php:2116]: Rank should not be prepend to the value code, values should be sorted then "output" in the correct order in the PHP definition like for other ranked elements.
- Simplifies logic and manipulations at runtime
- Easier maintainability

File: setup/compiler.class.inc.php [hunk 10]
```diff
@@ -2095,14 +2095,30 @@ protected function CompileAttribute(string $sAttType, DOMElement $oField, string
 			$this->CompileCommonProperty('allowed_values', $oField, $aParameters, $sModuleRelativeDir);
 			$aParameters['depends_on'] = $sDependencies;
 		} elseif ($sAttType == 'AttributeEnum') {
```
Comment (Molkobain) [inline setup/compiler.class.inc.php:2097]: Compilation missing for `AttributeMetaEnum`

File: core/valuesetdef.class.inc.php
Comment (Molkobain) [inline core/valuesetdef.class.inc.php:1]: See comment in compiler

File: core/attributedef.class.inc.php [hunk 11]
```diff
@@ -5504,6 +5504,7 @@ public function GetFormat()
 class AttributeEnum extends AttributeString
 {
 	const SEARCH_WIDGET_TYPE = self::SEARCH_WIDGET_TYPE_ENUM;
+	public $m_sSortOrder = 'code'; // Encryption library used for all encrypted fields
```
Comment (Molkobain) [inline core/attributedef.class.inc.php:5507]: What's that comment? Please verify what you copy/paste.

File: setup/compiler.class.inc.php [hunk 12]
```diff
@@ -2095,14 +2095,30 @@ protected function CompileAttribute(string $sAttType, DOMElement $oField, string
...
+			$oSortOrder = $oField->GetOptionalElement('sort_order');
+			if ($oSortOrder != null) {
+				$aParameters['sort_order'] = "'".$oField->GetChildText('sort_order')."'";
+			} else {
+				$aParameters['sort_order'] = "'code'";
```
Comment (Molkobain) [inline setup/compiler.class.inc.php:2102]: Why the single quotes within the double quotes, doesn't seem necessary. Single quotes are only necessary when you are about to print compiled code.

**Review by Molkobain — CHANGES_REQUESTED** [review]

**Comment by Molkobain** [comment]
Closing this PR as we need to correct it and merge it quickly. I'll handle it. Anne-Catherine might not have time to rework it as she is on support.

**Pull request closed without merging**